aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,138 @@
+ */
+
+#ifdef __KERNEL__
+#include <linux/string.h>
+#else
+#include <string.h>
+#endif
+
+#include "aesd-circular-buffer.h"
+#include <stdio.h>
+
+/**
+ * Helper function to advance the in_offs or out_offs pointers
+ * @param pointer: The 8-bit pointer to increment for read and write operations
+ * @return The incremented pointer. The pointer is wrapped around to the front of the circular buffer if the pointer has reached the end of the 
+ *         buffer
+ */
+uint8_t move_pointer(uint8_t cb_ptr)
+{
+    // If the pointer is at the last spot in the circular buffer
+    if (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED  == (cb_ptr + 1)){
+        return 0;
+    }
+    // Increment the pointer and return the updated value
+    else{
+        return (cb_ptr + 1);
+    }
+}
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    /**
+    * TODO: implement per description
+    */
+
+   // Error check 1: If the circular buffer is empty
+   if ((buffer->in_offs == buffer->out_offs) && (!buffer->full)){
+    return NULL;
+   }
+
+   // Local variable to store the resultant bytes travelled in the cirular buffer
+   size_t res_size_buffer = 0;
+   uint8_t max_writes_buffer = 0;
+
+   // Calculate the total number of bytes in the circular buffer
+   while(max_writes_buffer < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED){
+    res_size_buffer += buffer->entry[max_writes_buffer].size;
+    max_writes_buffer++;
+   }
+
+   // Error check 2: If the requested offset is greater than the actual number of bytes in the buffer
+   if (res_size_buffer < char_offset){
+    return NULL;
+   }
+
+    // Temporary variable to hold the current outoffs value
+    // First place to start reading from
+    uint8_t rptr = buffer->out_offs;
+
+    // Variable to hold the bytes in the circular buffer at the beginning of each iteration and at the end as well
+    // The second variable is to hold the number of bytes till the previous iteration to calculate the entry_offset_byte_rtn value
+    size_t curr_cb_count = 0 , prev_cb_count = 0;
+    
+
+    for (int i = 0; i < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED; i++){
```

Comment:
@KhyatiSatta this works but you could probably simplify and remove the additional loop above by just making this one loop handle the case where the total size of all elements are less than the offset requested.